Add Docker and compose

nicolasleger opened the pull request:
See `README` instructions to test

# Still To Do

- [x] MongoDB init
- [x] Fix error(s) at application launch

nicolasleger commented:
@adrienjoly I have this error when launching the Docker, any thoughts?

```
web_1    | npm info it worked if it ends with ok
web_1    | npm info using npm@5.0.3
web_1    | npm info using node@v8.1.3
web_1    | npm info lifecycle openwhyd@0.1.0~prestart: openwhyd@0.1.0
web_1    | npm info lifecycle openwhyd@0.1.0~start: openwhyd@0.1.0
web_1    |
web_1    | > openwhyd@0.1.0 start /usr/src/app
web_1    | > forever --outFile appd.out --errFile appd.err --eventFile app.events.js --minUptime 1000 --spinSleepTime 500 --number 100000 start app.js $@
web_1    |
web_1    | info:    Forever processing file: app.js
web_1    | npm info lifecycle openwhyd@0.1.0~poststart: openwhyd@0.1.0
web_1    | npm info ok
openwhyd_web_1 exited with code 0
```

adrienjoly commented:
Coucou Nicolas! Thanks for your pull request! :-)

I don't see what the error is, but here are some things you may want to try:

- make sure that environment variables are set (e.g. with `env-vars-testing.sh`)
- make sure that mongodb instance is running in the container, and initialized with `initdb.js`
- use `npm run run` (foreground process => it will display logs and errors directly in stdout and stderr) instead of `npm start` (daemon logging to a file, using "forever")

These files may provide some guidance:

- https://github.com/openwhyd/openwhyd/blob/master/README.md
- https://github.com/openwhyd/openwhyd/blob/master/.travis.yml

adrienjoly commented:
(for reference, associated issue: https://github.com/openwhyd/openwhyd/issues/56)

adrienjoly commented:
are you still on this, @nicolasleger? or do you want someone else to continue from this stage?

nicolasleger commented:
Thanks a lot @adrienjoly. I'm still on it for lunch, hope to fix errors. 🤞

nicolasleger commented:
@adrienjoly I moved forward with your help. Now on a `mongo` or `mongoose` module error.

```
web_1    | > openwhyd@0.1.0 run /usr/src/app
web_1    | > node app.js --color --fakeEmail --digestInterval -1 $@
web_1    |
web_1    | { Error: Cannot find module '../build/Release/bson'
web_1    |     at Function.Module._resolveFilename (module.js:485:15)
web_1    |     at Function.Module._load (module.js:437:25)
web_1    |     at Module.require (module.js:513:17)
web_1    |     at require (internal/module.js:11:18)
web_1    |     at Object.<anonymous> (/usr/src/app/node_modules/bson/ext/index.js:15:10)
web_1    |     at Module._compile (module.js:569:30)
web_1    |     at Object.Module._extensions..js (module.js:580:10)
web_1    |     at Module.load (module.js:503:32)
web_1    |     at tryModuleLoad (module.js:466:12)
web_1    |     at Function.Module._load (module.js:458:3) code: 'MODULE_NOT_FOUND' }
web_1    | js-bson: Failed to load c++ bson extension, using pure JS version
```

Might help: https://stackoverflow.com/questions/28651028/cannot-find-module-build-release-bson-code-module-not-found-js-bson

adrienjoly commented:
This warning should not prevent the app server from running. So I guess that you're getting another error, besides this warning.

adrienjoly commented:
please let me know if you want someone else to have a look at this, or if you prefer to keep working solo on this PR :-)

nicolasleger commented:
I updated and fixed Docker config however I got an empty response at startup.
The line `mongodb.MongoClient.connect(url, options, function(err, db){` does not trigger anything despite seeding the database before. I got the same locally without Docker and proper initialized database.

adrienjoly commented:
you mean that `connect()` does not call the callback function?

that would explain why openwhyd/whydjs server hangs at startup, on claire's machine! (@cmdcmdcmd) 

any clue on how to fix this?

nicolasleger commented:
Yes It looks like the same problem.
Might come from the npm `mongodb` module or MongoDB version, or node ?.
Tested with MongoDB `v3.2.11` / `v3.4.6` and `mongodb@1.4.40` for module.

adrienjoly commented:
Enabling acknowledgments more explicitly could solve the problem, cf https://mongodb.github.io/node-mongodb-native/driver-articles/mongoclient.html. (see the `w` parameter)

nicolasleger commented:
Thanks, but just tested `?w=1` without success.

nicolasleger commented:
What are your MongoDB, `node`, `npm`, `bson` and `nan` modules versions please?

adrienjoly commented:
```
$ node --version
v6.6.0
$ npm --version
3.10.3
$ npm ls | grep bson
│ ├─┬ bson@0.2.22
$ npm ls | grep nan
│ │ │ │ └── number-is-nan@1.0.1
│ │ │ │   │ │ │   └── number-is-nan@1.0.1
│ └── nan@2.6.2
│ │ └── nan@1.8.4
│ │ └── nan@1.8.4
```

adrienjoly commented:
maybe we should match these versions in the container, create a separate issue for the mongodb connect callback problem, then update the container after we solve that problem?

nicolasleger commented:
I agree, I will try to setup the container with those versions and check if ok.

adrienjoly commented:
@nicolasleger ... or use the ones that are currently running in production:

```bash
$ mongod --version
db version v3.2.13
git version: 23899209cad60aaafe114f6aea6cb83025ff51bc

$ node --version
v4.8.2

$ npm --version
2.15.11

$ npm ls | grep bson
│ ├─┬ bson@0.2.22

$ npm ls | grep nan
│ │ │ │ └── number-is-nan@1.0.1
│ │ │   └── number-is-nan@1.0.1
│ └── nan@2.6.2
│ │ └── nan@1.8.4
│ │ └── nan@1.8.4
```

nicolasleger commented:
Indeed it works now with `production` versions, thanks!
Can you please test it on your side following `README` instructions relative to Docker?

adrienjoly commented:
Awesome! I'm on it!

In the mean time, would you mind opening a new issue for the "mongodb connect callback" problem? (if so, don't forget to mention the versions that did not work)

adrienjoly commented on README.md:
```diff
@@ -54,7 +54,16 @@ Thank you for your understanding! ^^
-## Setup
+## Setup (with Docker, for development environment)
+
+* Instal Docker Client via https://www.docker.com/
```
would it be too specific to use that URL instead: https://www.docker.com/community-edition ?
(as the first download link I found on the homepage was for the entreprise edition, and I believe that the CE is enough.)

adrienjoly commented on docker-compose.yml:
```diff
@@ -0,0 +1,45 @@
+    depends_on:
+      - mongo
+    environment:
+      # From whydJS/env-vars-testing.sh not working in Dockerfile
```
did you try using the `source whydJS/env-vars-testing.sh` ?

nicolasleger replied:
Yes but docker manage ENV on its own and I didn't succeed. Also you have to keep here at least `MONGODB_HOST=mongo` and not redefining it in script.

adrienjoly replied:
alright, that's ok for now :-)

nicolasleger commented:
Issue created at https://github.com/openwhyd/openwhyd/issues/58

adrienjoly commented:
when running `docker-compose up` for the first time after, I'm getting this:

```
web_1    | /usr/src/app/node_modules/mongodb/lib/mongodb/connection/base.js:246
web_1    |         throw message;      
web_1    |         ^
web_1    | 
web_1    | TypeError: Cannot read property 'find' of undefined
web_1    |     at Object.exports.fetchMulti (/usr/src/app/app/models/user.js:198:29)
web_1    |     at Object.exports.cacheUsers (/usr/src/app/app/models/mongodb.js:79:12)
web_1    |     at finishInit (/usr/src/app/app/models/mongodb.js:171:12)
web_1    |     at /usr/src/app/app/models/mongodb.js:179:6
web_1    |     at /usr/src/app/node_modules/mongodb/lib/mongodb/db.js:572:5
web_1    |     at /usr/src/app/node_modules/mongodb/lib/mongodb/command_cursor.js:164:18
web_1    |     at /usr/src/app/node_modules/mongodb/lib/mongodb/db.js:1197:7
web_1    |     at /usr/src/app/node_modules/mongodb/lib/mongodb/db.js:1905:9
web_1    |     at Server.Base._callHandler (/usr/src/app/node_modules/mongodb/lib/mongodb/connection/base.js:453:41)
web_1    |     at /usr/src/app/node_modules/mongodb/lib/mongodb/connection/server.js:488:18
mongo_1  | 2017-07-12T14:41:31.419+0000 I NETWORK  [conn6] end connection 172.18.0.3:46378 (4 connections now open)
mongo_1  | 2017-07-12T14:41:31.419+0000 I NETWORK  [conn5] end connection 172.18.0.3:46376 (3 connections now open)
mongo_1  | 2017-07-12T14:41:31.420+0000 I NETWORK  [conn4] end connection 172.18.0.3:46374 (2 connections now open)
mongo_1  | 2017-07-12T14:41:31.420+0000 I NETWORK  [conn3] end connection 172.18.0.3:46372 (1 connection now open)
mongo_1  | 2017-07-12T14:41:31.420+0000 I NETWORK  [conn2] end connection 172.18.0.3:46370 (0 connections now open)
web_1    | 
web_1    | npm info openwhyd@0.1.0 Failed to exec run script
web_1    | npm ERR! Linux 4.9.31-moby
web_1    | npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "run"
web_1    | npm ERR! node v4.8.4
web_1    | npm ERR! npm  v2.15.11
web_1    | npm ERR! code ELIFECYCLE
web_1    | npm ERR! openwhyd@0.1.0 run: `node app.js --color --fakeEmail --digestInterval -1 $@`
web_1    | npm ERR! Exit status 1
web_1    | npm ERR! 
web_1    | npm ERR! Failed at the openwhyd@0.1.0 run script 'node app.js --color --fakeEmail --digestInterval -1 $@'.
```

=> I guess that it's normal because I didn't run the db init script yet. Wouldn't it make sense to include these scripts in the docker-compose file to prevent this error?

adrienjoly commented:
(I'm a docker noob, FYI)

... after the error explained above, I had to run the `docker-compose exec mongo mongo openwhyd_data data/initdb.js data/initdb_team.js && docker-compose restart web` command in a different terminal.

(I realised that pressing ctrl-c in the terminal that displays the error actually stops the container. so I had to restart it before running the db init command.)

=> should we include that in the setup procedure of README.md, in your opinion?

adrienjoly commented:
all that being said, it works! => great job, nicolas!

nicolasleger commented:
That's right following the `README` note you have to execute it just after the first build.

Not sure how to include it in `docker-compose` because `command` will have to launch and execute the script.
